src/status_im/chat/events/sign_up.cljs

```diff
@@ -49,37 +49,85 @@
+ (fn [{:keys [db]}]
+   {:db                   (dissoc db :confirmation-code-sms-listener)
+    ::remove-sms-listener (:confirmation-code-sms-listener db)}))
+
+(re-frame/reg-fx
+ ::remove-sms-listener
+ (fn [subscription]
+   (sms-listener/remove-sms-listener subscription)))
+
+(defn- sms-receive-handler [{confirmation-code :body}]
+  (when-let [matches (re-matches #"(\d{4})" confirmation-code)]
+    (re-frame/dispatch [::sign-up-confirm (second matches)])))
+
+(def start-listening-confirmation-code-sms-event
+  [:request-permissions
+   [:receive-sms]
+   (fn []
+     (let [listener (sms-listener/add-sms-listener sms-receive-handler)]
+       (re-frame/dispatch [:start-listening-confirmation-code-sms listener])))])
+
+(handlers/register-handler-fx
+ ::sign-up-success
+ [re-frame/trim-v (re-frame/inject-cofx :random-id)]
+ (fn [{:keys [db random-id]} message-id]
+   (-> {:db         db
+        :dispatch-n [;; create manual way for entering confirmation code
+                     (sign-up/enter-confirmation-code-event random-id)
+                     ;; create automatic way for receiving confirmation code
+                     start-listening-confirmation-code-sms-event]}
+       (message-seen message-id))))
+
+(handlers/register-handler-fx
+ :update-phone-number-in-profile
+ [re-frame/trim-v]
+ (fn [{{:keys [chats] :as db} :db}]
+   (let [phone-command (->> (get-in chats ["console" :messages])
+                            (filter (fn [{:keys [type content]}]
+                                      (and (= type :response)
+                                           (= (:command content) "phone"))))
+                            (first))]
+     {:db         db
+      :dispatch-n [[:account-update {:phone (get-in phone-command [:content :params :phone])}]]})))
+
+(defn- extract-last-phone-number [chats]
+  (let [phone-message (->> (get-in chats ["console" :messages])
+                           (filter (fn [{:keys [type content]}]
+                                     (and (= type :response)
+                                          (= (:command content) "phone"))))
+                           (first))]
+    (get-in phone-message [:content :params :phone])))
 
 (handlers/register-handler-fx
   ::sign-up-confirm
   (fn [{:keys [db]} [confirmation-code message-id]]
     (sign-up-confirm db confirmation-code message-id)))
 
 (defn- sign-up-confirmed [{:keys [db] :as fx}]
-  (cond-> (update fx :dispatch-n conj [:request-permissions
-                                       [:read-contacts]
-                                       #(re-frame/dispatch [:sync-contacts (fn [contacts]
-                                                                             [::contacts-synced contacts])])])
-    (:confirmation-code-sms-listener db)
-    (merge {:db                  (dissoc db :confirmation-code-sms-listener)
-            :remove-sms-listener (:confirmation-code-sms-listener db)})))
+  (let [last-phone-number (extract-last-phone-number (:chats db))]
```

Comment:
It doesn't feel right to filter through *all* messages in particular chat (even if it's console) just to get the parameter of event which started the whole process (sign-up).
It's even prone to errors, what if I issued the `/phone` command multiple times and this filtering will incorrectly select the phone number from the first command message it found, while there is some later `/phone` command message ?

Reply:
Actually, you're wrong. The last message is the first in the :messages list, so it's okay to just pick the first `/phone` in this list. 
Agree that it's not wise to use `filter` here because it is not efficient, but the problem will be gone if I replace it with `some`.

Reply:
Wouldn't it be better to just have a phone number available in the `/confirmation-code` command-request, so we don't have to do any search for it ?
The `/confirmation-code` command-request is generated as a response to the `/phone` command with particular phone number, so it seems logical and correct to do it that way.